### PR TITLE
[FEATURE] Improved and more hooks to the project

### PR DESCRIPTION
**Pull request by mk-mxp**
I wanted to use the existing hooks (some time ago) and found them to be cluttered and incomplete. So I wrote a hook service class and added it to the DefaultController where I needed hooks.

It is based on the way the (most recently added?) hooks where already implemented.

Please discuss these issues with my implementation:

- Where to place the interfaces? PHP does not allow "interface" in the namespace, so using PSR-4 required some other place.
- Testing the hook service required a test hook interface. Where to place that?
- Are the hooks complete enough? Using interfaces makes it harder to add more places / methods to a hook (becoming backwards incompatible immidiately).

The tests for the hooks are still missing. I got caught by You switching to nimut and need to get the test run locally again before adding more tests...

### CODE REVIEW

**Comment by oliverklee** [comment]
> I wanted to use the existing hooks (some time ago) and found them to be cluttered and incomplete.

I agree. :-)

@mk-mxp Thanks for taking care of this! I'll review it during the next few days.

**Comment by mk-mxp** [comment]
Tests added.

**Comment by mk-mxp** [comment]
How about this? I have some more hooks implemented the same way, need the basics clarified before going on.

**Comment by oliverklee** [comment]
@mk-mxp Thanks a lot for the PR! I really appreciate it. It might take a few days until I can provide feedback to the concept as I'm really busy this month. Please bear with me in the meantime.

**Comment by mk-mxp** [comment]
> @mk-mxp Thanks a lot for the PR! I really appreciate it. It might take a few days until I can provide feedback to the concept as I'm really busy this month. Please bear with me in the meantime.

Ping.

**Comment by oliverklee** [comment]
@mk-mxp Thanks for the ping. I'm awfully sorry for the delay. I'm currently working on getting oelib/onetimeaccount/seminars compatible with TYPO3 9LTS (which is the crowdfunded project) and will provide feedback for your work directly after that.

**Comment by mk-mxp** [comment]
How big are chances that his will fit into the upcoming V3 (Typo3 V9 compatible) without major changes?

**Comment by oliverklee** [comment]
Good question. Okay, I'll look at your changes in time before the 3.0 release so we can get any breaking changes in before the release.

File: Classes/BagBuilder/Abstract.php [hunk 1]
```diff
@@ -169,6 +169,49 @@ public function getWhereClause()
...
+     * Returns a WHERE clause part for the bag to create.
+     *
+     * If the bag does not have such limitation imposed upon, the return value
+     * will be empty.
+     *
+     * @param string $key the limitation key to return, must not be empty
+     *
+     * @return string WHERE clause part for the bag to create, may be empty
+     */
+    public function getWhereClausePart($key)
```
Comment (oliverklee) [inline Classes/BagBuilder/Abstract.php]: I like this change. Please let's have it in a separate PR.

File: Classes/FrontEnd/DefaultController.php [hunk 2]
```diff
@@ -3047,9 +3150,27 @@ protected function createRegistrationForm()
...
+
         return $registrationEditor->render();
     }
 
+    /**
+     * Creates the registration page footer.
+     *
+     * @return string HTML code including the title and error message
+     */
+    protected function createRegistrationFooter()
```
Comment (oliverklee) [inline Classes/FrontEnd/DefaultController.php]: Please let's move this refactoring (without the hook at first) into a separate PR.

File: Classes/Interfaces/Hook.php [hunk 3]
```diff
@@ -0,0 +1,13 @@
+<?php
+
+namespace OliverKlee\Seminars\Interfaces;
+
+/**
+ * Extend this interface to use Your interface with the HookService
```
Comment (oliverklee) [inline Classes/Interfaces/Hook.php]: Extend -> implement

File: Classes/Interfaces/Hook/RegistrationView.php [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+namespace OliverKlee\Seminars\Interfaces\Hook;
+
+use OliverKlee\Seminars\Interfaces\Hook;
+
+/**
+ * This interface needs to be used for hooks concerning the registration view.
+ *
+ * @author Michael Kramer <m.kramer@mxp.de>
+ */
+interface RegistrationView extends Hook
```
Comment (oliverklee) [inline Classes/Interfaces/Hook/RegistrationView.php]: There is no registration view. Maybe we can call it `RegistrationForm`?

Comment (mk-mxp) [reply Classes/Interfaces/Hook/RegistrationView.php]: Actually, yes, it is the registration form. In MVC terms this is hooking into the View part of the registration form. So using "View" is consistent with the other view types (what_to_display in FlexForms). Is RegistrationFormView OK?

Comment (oliverklee) [reply Classes/Interfaces/Hook/RegistrationView.php]: The `RegistrationForm` class is a bad mashup of a controller and a view (and more of a controller than a view). I'd prefer to call it `RegistrationForm` as long as the registration form has not been migrated to (i.e., rewritten in) Extbase/Fluid.

File: Classes/BagBuilder/Abstract.php (hunk 1 above)
Comment (mk-mxp) [reply Classes/BagBuilder/Abstract.php]: Resolved by: #308

File: Classes/FrontEnd/DefaultController.php (hunk 2 above)
Comment (mk-mxp) [reply Classes/FrontEnd/DefaultController.php]: Resolved by #309

File: Classes/Interfaces/Hook.php (hunk 3 above)
Comment (mk-mxp) [reply Classes/Interfaces/Hook.php]: This interface is NOT for implementation, it is the [base for other hook interfaces only](https://github.com/oliverklee/ext-seminars/blob/48e50e10472da840c6e6613973aa704e0b603d41/Classes/Interfaces/Hook/SeminarListView.php#L16).

Comment (oliverklee) [reply Classes/Interfaces/Hook.php]: Ah, okay. Then it make sense. :-) Thanks for the explanation!

**Comment by mk-mxp** [comment]
All functionality of this was split up to several PRs and this is obsolete now.

**Pull request closed without merging**